### ISSUE TITLE
[BUG] - adding collections sometimes modifies the originals

### ISSUE THREAD

**Issue by elfehr** [magpylib/magpylib#472]
After creating a bigger collection made of several collections (instead of adding all magnets individually), calculating the B field of the first collection returns the B field of the bigger collection. (version 3.0.2)

**To Reproduce**
I define a collection c1 of two magnets and get its B field normally. But if I then create a bigger collection from c2 = c1 + some other magnet(s) and calculate again the B field of c1, I get the B field of c2 instead.

The H field is not affected. Redefining c1 before recalculating the B field gives the normal field again.
The problem doesn't occur if I add up the original magnets instead of adding up collections, or if I add the magnet first + the collection last. But it also doesn't occur if I'm calculating the B field of the Box collection instead of the Cylinder collection, so I'm not quite sure what is going on.

```python
import numpy as np
import matplotlib.pyplot as plt
import magpylib as magpy
from scipy.spatial.transform import Rotation as R

# metals
m1 = magpy.magnet.Box(magnetization=(0, 1, 0),
                       dimension=(10, 10, 3.5e-6),
                       position=(0, 0, 0))
m2 = magpy.magnet.Box(magnetization=(1, 0, 0),
                       dimension=(10, 10, 3.5e-6),
                       position=(0, 0, 0))

# electromagnet
em1 = magpy.magnet.Cylinder(magnetization=(0, 0, 4),
                            dimension=(30, 10),
                            position=(-15, 0, 0),
                            orientation=R.from_euler('y', 90, degrees=True))
em2 = magpy.magnet.Cylinder(magnetization=(0, 0, 4),
                            dimension=(30, 10),
                            position=(15, 0, 0),
                            orientation=R.from_euler('y', 90, degrees=True))

# prepare xy grid at z=0
dim = 10
n = 20 # points
xs = np.linspace(-dim, dim, n)
ys = np.linspace(-dim, dim, n)
loc = np.array([(x, y, 0) for y in ys for x in xs])

# fields generated by collections
c0 = m1 + m2
c1 = em1 + em2
A = c1.getB(loc).reshape(n, n, 3)

c2 = c1 + c0 # <----- causes the problem
# c2 = c1 + m1 # <----- causes the problem
# c2 = c0 + c1 # <----- doesn't cause the problem
# c2 = m1 + c1 # <----- doesn't cause the problem
# c2 = em1 + em2 + m1 + m2 # <----- doesn't cause the problem
# c1 = em1 + em2 # <----- redefining c1 solves the problem
B = c1.getB(loc).reshape(n, n, 3)

# fields components
X, Y = np.meshgrid(xs, ys)
Ax, Ay = A[:, :, 0], A[:, :, 1]
Bx, By = B[:, :, 0], B[:, :, 1]

# plot
fig, (ax1, ax2) = plt.subplots(1, 2)
ax1.streamplot(X, Y, Ax, Ay)
ax2.streamplot(X, Y, Bx, By)
ax1.set_title('B(c1) before')
ax2.set_title('B(c1) after')
plt.show()


```

**Expected behavior**
I expected that creating a collection by adding smaller collections would be equivalent to adding all the magnets separately.

**Screenshots**
![Figure_1](https://user-images.githubusercontent.com/86655195/153273789-d82a6b4b-139a-40b0-aa16-2e71ab78d8f7.png)

**Desktop (please complete the following information):**
 - OS: Linux 64bit (Manjaro)
 - IDE: running the above script with Python 3.10.2

**pip freeze output**

```
pip freeze output:
anytree==2.8.0
apipkg==2.1.0
apparmor==3.0.3
appdirs==1.4.4
application-utility==1.3.2
apsw==3.36.0.post1
arandr==0.1.10
argcomplete==1.12.3
argon2-cffi==21.1.0
asciidoc==10.1.1
asttokens==2.0.5
atomicwrites==1.4.0
attrs==21.4.0
Automat==20.2.0
backcall==0.2.0
bauh==0.9.26
bcrypt==3.2.0
Beaker==1.11.0
beautifulsoup4==4.10.0
black==22.1.0
borgbackup==1.1.17
bottle==0.12.19
brotlicffi==1.0.9.2
bsddb3==6.2.9
btrfsutil==5.16
CacheControl==0.12.6
cchardet==2.1.7
certifi==2021.10.8
cffi==1.15.0
chardet==4.0.0
click==8.0.3
click-log==0.3.2
click-threading==0.5.0
cmasher==1.6.3
colorama==0.4.4
colorcet==3.0.0
colorspacious==1.1.2
conan==1.45.0
ConfigArgParse==1.5
configobj==5.1.0.dev0
constantly==15.1.0
construct==2.10.67
contextlib2==0.6.0.post1
cryptography==36.0.1
css-parser==1.0.6
cssselect==1.1.0
cycler==0.11.0
decorator==5.1.1
deluge==2.0.5
deprecation==2.1.0
discid==1.2.0
distlib==0.3.4
distro==1.6.0
dnspython==2.2.0
e13tools==0.9.6
executing==0.8.2
extras==1.0.0
fasteners==0.16.3
feedparser==6.0.2
fixtures==3.0.0
fonttools==4.28.5
future==0.18.2
fuzzywuzzy==0.18.0
GeoIP==1.3.2
gitdb==4.0.7
GitPython==3.1.24
gpg==1.16.0
gufw==21.4.0
hjson==3.0.2
html2text==2020.1.16
html5-parser==0.4.10
html5lib==1.1
httplib2==0.20.2
hyperlink==21.0.0
icalendar==4.0.9
idna==3.3
ifaddr==0.1.7
importlib-metadata==4.8.1
incremental==21.3.0
iniconfig==1.1.1
ipython==8.0.1
ipython_genutils==0.2.0
jedi==0.18.1
jeepney==0.7.1
Jinja2==3.0.3
khal==0.10.4
kiwisolver==1.3.2
LibAppArmor==3.0.3
libtorrent===2.0.5-build-libtorrent-rasterbar-src-libtorrent-rasterbar-2.0.5-bindings-python
lit==13.0.0.dev0
llfuse==1.4.1
louis==3.20.0
lxml==4.7.1
magpylib==3.0.2
Mako==1.1.6
manjaro-sdk==0.1
Markdown==3.3.6
MarkupSafe==2.0.1
matplotlib==3.5.1
matplotlib-inline==0.1.3
mechanize==0.4.6
meson==0.61.1
monotonic==1.6
more-itertools==8.10.0
mox3==1.1.0
msgpack==1.0.3
multivolumefile==0.2.3
mutagen==1.45.1
mypy-extensions==0.4.3
netifaces==0.11.0
netsnmp-python==1.0a1
node-semver==0.8.0
npyscreen==4.10.5
numpy==1.22.0
oauthlib==3.1.1
ordered-set==4.0.2
packaging==20.9
pacman-mirrors==4.23.2
pandas==1.3.5
param==1.12.0
paramiko==2.9.2
parso==0.8.2
patch-ng==1.17.4
pathspec==0.9.0
pbr==5.8.0
pdfannots==0.3
pdfminer.six==20211012
pep517==0.12.0
pexpect==4.8.0
picard==2.7.3
pickleshare==0.7.5
Pillow==9.0.1
platformdirs==2.4.1
pluggy==0.13.1
pluginbase==1.0.0
ply==3.11
progress==1.6
prompt-toolkit==3.0.26
psutil==5.9.0
ptyprocess==0.7.0
pure-eval==0.2.2
py==1.11.0
py7zr==0.16.2
pyaml==21.8.3
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycairo==1.20.1
pychm==0.8.6
pycparser==2.21
pycryptodome==3.14.0
pycryptodomex==3.10.1
pyct==0.4.8
pygame==2.1.2
Pygments==2.11.2
PyGObject==3.42.0
PyHamcrest==1.10.1
PyJWT==2.2.0
pykeepass==4.0.1
PyNaCl==1.5.0
pyOpenSSL==21.0.0
pyparsing==2.4.7
pyppmd==0.17.1
PyQt4_sip==4.19.25
PyQt5==5.15.6
PyQt5-sip==12.9.1
PyQtWebEngine==5.15.5
PySide6==6.2.3
pysmbc==1.0.23
PySocks==1.7.1
pytest==6.2.5
python-dateutil==2.8.2
python-Levenshtein==0.12.2
python-slugify==5.0.2
python-xlib==0.31
pytz==2021.3
pytz-deprecation-shim==0.1.0.post0
pywal==3.3.0
pyxdg==0.27
PyYAML==5.4.1
pyzstd==0.15.0
regex==2021.11.10
rencode==1.0.6
reportlab==3.6.6
requests==2.26.0
requests-oauthlib==1.3.0
requests-toolbelt==0.9.1
resolvelib==0.5.5
retrying==1.3.3
rofimoji==5.3.0
ruamel.yaml==0.17.20
ruamel.yaml.clib==0.2.6
scipy==1.7.3
service-identity==21.1.0
setproctitle==1.2.2
sgmllib3k==1.0.0
shiboken6==6.2.3
shtab==1.5.3
six==1.16.0
smmap==5.0.0
soupsieve==2.2.1
speedtest-cli==2.1.3
sshtunnel==0.4.0
stack-data==0.1.4
TBB==0.2
team==1.0
termcolor==1.1.0
testtools==2.5.0
text-unidecode==1.3
texttable==1.6.4
tldr==3.0.0
toml==0.10.2
tomli==2.0.0
torbrowser-launcher==0.3.5
tqdm==4.62.3
traitlets==5.1.1
Twisted==21.7.0
typing_extensions==4.0.1
tzdata==2021.5
tzlocal==4.0.1
ufw==0.36.1
unrardll==0.1.4
urllib3==1.26.8
urwid==2.1.2
validate==5.1.0.dev0
vdirsyncer==0.18.1.dev0+g3191886.d20211204
wcwidth==0.2.5
webencodings==0.5.1
websockets==10.1
yt-dlp==2022.1.21
zeroconf==0.38.1
zipp==3.7.0
zope.interface==5.4.0
```

**Comment by OrtnerMichael** [comment]
hi @elfehr,

Thank you for bringing this to our attention !

In the new version 4 this problem does not seem to persist anymore. This is what I get when I run your code:
![image](https://user-images.githubusercontent.com/47384012/153373576-bde086c4-92bf-48f5-82c7-61c4f680e655.png)

Version 4 is complete (api-wise) - but it is not published yet. We are just fixing the docstrings, which should take about a week. You can already use Version 4 by downloading it from the [main branch](https://github.com/magpylib/magpylib). We are still using source distribution, so you do not even need to install it (but you can with setup.py), but just copy the magpylib (package) folder into your working directory.

I'm not sure if we will continue to update version 3 - at the moment we are fully occupied with releasing v4.

There is quite a lot of refactoring in version 4. E.g. Box is now names Cuboid, display is now called show.

sincerely,
michael

**Comment by elfehr** [comment]
Hi, thanks for the answer!

That's great if the bug isn't there anymore in the new version. I'll try it a one point.

I understand trying to track this bug wouldn't make much sense currently — at least if anybody manages to reproduce it, they have a chance to end up finding a solution here.

**Comment by OrtnerMichael** [comment]
hi @elfehr 

We just deployed a pre-release of version 4 to PyPI.

You can install it with
```
pip install magpylib==4.0.0-beta2
```

In your code you also have to replace `Box` with `Cuboid`. The rest should be fine.

sincerely,
michael

**Comment by elfehr** [comment]
Great, thanks! That's easier.

I can confirm the same code (except replacing Box with Cuboid) shows the bug in 3.0.2 and not in the pre-release version.

**Comment by OrtnerMichael** [comment]
Thank you for the feedback !
Let us know if you need some assistance.
